### ISSUE TITLE
Image::row() with negative strides

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#815]
The code [here](https://github.com/MRtrix3/mrtrix3/blob/master/lib/image.h#L97-L104) obtaining a map for accessing all data along a particular axis of an image (e.g. data for all volumes for a particular voxel) fails an assertion in Eigen 3.2.93 if the axis of interest has a negative stride.

So we need to look into whether there's a way to 'trick' Eigen so that this functionality can still be used, or perhaps whether we can provide our own map wrapper class (albeit perhaps losing Eigen functionality / compatibility).

**Comment by jdtournier** [comment]
I've been thinking about our increasingly heavy usable of the `row()` method and some of the problems it introduces. I think we need to rethink how it's implemented and used in the code, as things stand it's too fragile, and not necessarily as efficient as it might seem. 

- The first thing is that the image needs to be access with direct IO, which can already introduce issues when developers forget to specify this. It's easy to overlook in testing since MRtrix will try very hard to memory-map its input it at all possible, so will basically access the image with direct IO if it can. And our apps are typically geared to producing outputs that are suitable for direct mapping in downstream apps (e.g. `dwi2fod` will produce output with volume-consecutive strides in `Float32` format, which is what `tckgen` will expect). But things will fall over the minute someone else feeds in data that isn't directly suitable for it. It's easy to fix by adding the `.with_direct_IO()` call, but it feels wrong to have such a low-level call be so dependent on how the image was accessed in the first place. And also forces preloads in cases where they might not have been necessary (e.g. because the data type doesn't match, but it's still a single-pass application). 

- the way it's implemented, we have to set the returned vector (or rather `Eigen::Map`) as `Eigen::Unaligned`. This means it's _not_ suitable for vectorisation. So any performance-critical section that uses this vector (i.e. just about anything we do with it) will either be several times slower (no vectorisation) or will involve an on-the-fly copy to registers (an overhead that would hinder performance it the vector is used in multiple operations). 

- and finally, as @Lestropie points out, it's not even guaranteed to work properly when the relevant stride is negative. _MRtrix3_ won't bother preloading if the strides match _ignoring their sign_ - this is because performance won't be affected by this, the CPU's memory access prediction engine will be perfectly capable of dealing with this, and cache usage remains just as efficient. We could force a preload in these cases, but this feels a bit OTT, especially given all the concerns above. 

So what I suggest is that we change the implementation to something that returns a _copy_ of the data, which will guarantee alignment, allow usage with non-direct IO data or data with negative strides, and in the worst case scenario won't affect performance (since a copy will be performed by Eigen anyway to allow vectorisation) or actively _boost_ performance if vectorisation wasn't being used before. 

To retain the convenience of the method, there are a couple of options:

1. `.row()` returns a normal `Eigen::Matrix<value_type, Eigen:: Dynamic, 1>`, and we add another method `.row (Eigen::Vector<>& )` to set the values for those cases where it's required. We'd need to be a bit careful about handling the returned vector to avoid unnecessary copies and allocations, but otherwise doable.

2. `.row()` returns a proxy object that behaves a bit like `Eigen::Map`, but is designed to be allocated to or from an Eigen container (not necessarily `Eigen::VectorXd` or the like). This would give us maximum flexibility, code legibility and performance, but might require a bit of effort to set up. This would be my preferred option... 

Note that in the latter case, we won't be able to use constructs such as:

    auto values = image.row(3);

and expect `values` to behave like any other Eigen container, since its type will definitely _not_ be an Eigen object. You'd need to have an existing container (the preferred option for performance reasons) or declare the type upfront, e.g.:

    Eigen::VectorXd values;
    ...
    values = image.row (3);

or 

    Eigen:VectorXd values = image.row (3);

The latter involves an allocation, and depending on where the call is made, might result in reduced performance. It would be better to always allocate upfront (maybe as a class member) to allow reuse of the container. 

Note that option 1 will _always_ entail an allocation, which is why I'm not overly keen on it. 

Anyway, just my thoughts, we can discuss after ISMRM...

**Comment by jdtournier** [comment]
Just to add one more thought while it's fresh in my mind: 

To implement the above with the least overhead, we need calls have knowledge of both sides of the assignment. In other words, whatever does the copy needs to have access to the Eigen container that is being assigned to (so it can resize if required, perform type conversions, etc.), otherwise the only option is a return by copy, which brings in a memory allocation. Unfortunately this makes it impossible to implement a syntax like:

    Eigen::VectorXd values;
    values = image.row (3);

without an unnecessary allocation, since the `.row()` method can't access the `values` container. It can't adjust to a mismatch in the data type either, although that's unlikely to be much of a concern in practice. It might be possible for the `.row()` method to return a proxy class compatible with Eigen's `operator=()`, which would get around the issues altogether. But that would need a bit of investigating. 

Otherwise, the only operators that we could realistically overload to achieve this are `<<` and/or `>>` - which interestingly Eigen uses already for very much the same reasons. So this would look like:

    Eigen::VectorXd values;
    values << image.row (3);
    ...
    image.row (3) << values;

Only snag is it might interfere with [Eigen's existing equivalent initialisers](https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html)... In this case, we could opt for the opposite:

    Eigen::VectorXd values;
    image.row (3) >> values;
    ...
    values >> image.row (3);

Ideally we'd allow either, but that very much depends on whether the compiler can distinguish our overload from Eigen's.

Anyway, just a few thoughts while you guys slave away on your abstracts...

**Comment by draffelt** [comment]
> It might be possible for the .row() method to return a proxy class compatible with Eigen's operator=(), which would get around the issues altogether. But that would need a bit of investigating.

If this doesn't work, then I think I would prefer to use more standard syntax, even if it's slightly more verbose:
```
Eigen::VectorXd values;
image.row (values, 3)
```
And make sure `value` is declared outside the loop or is a class member.  It's still easier than manually looping over the desired axis to fill the vector.

**Comment by jdtournier** [comment]
Agreed that something like this is needed for convenience - it's a very common operation. Unfortunately I'm pretty sure the assignment operator isn't going to be feasible: it would have to hook deep into Eigen's backend code, and probably in a way that won't be compatible between versions (evaluation of operations in the assignment operator is one of the things they changed in Eigen 3.3). So I don't think it's a good idea.

Your suggestion of an `image.row (values, 3);` sounds OK at first, until you realise that can't tell a read from a write with that syntax... You could have a `.row()` and an equivalent `.save_row()` or something method, I suppose. 

But the one I'm favouring at the moment is the insertion/extraction operators. It would be pretty simple to have a `.row()` method return a proxy class with a reference to the image and the desired axis, and have global `operator<<` and `operator>>` methods defined to do the work. We already use a similar approach for the `.value()` and `.index()` methods, this would actually be even simpler. This would allow this kind of syntax:

```C++
Eigen::VectorXd values;
...
values << image_in.row(3);
// do stuff
image_out.row(3) << values;
```
and as you say, values would be declared as a class member to avoid unnecessary reallocations. But you would also be able to use it on any vector-like Eigen type. And we could also define the extraction operators to do the same thing (not strictly necessary, but why not). So for example, this would be entirely possible:
```C++
Eigen::MatrixXd data_in;
Eigen::VectorXd data_out;
...
image1.row(3) >> data_in.col(0);
image2.row(3) >> data_in.col(1);
image3.row(3) >> data_in.col(2);
// do stuff
data_out >> image_out.row(3);
```
These operators are as far as I can tell the only ones that give us that flexibility - all the others are declared at class scope, and expect the class to be the LHS operand (which is why it's tricky to use the assignment operator for the data read - wouldn't be a problem for the write-back, since the assignment would then be a member of the image class). The only other suitable operator which can be declared outside class scope is the comma operator - but who really uses that?!? Eigen does for their initialiser code (that's how you can write `Eigen::VectorXd vec(3) << 1, 2, 3;`), but I have to say I find that pretty unusual... Besides, the extraction / insertion operators have an implied meaning that's actually quite a good match here.

What do you reckon?

**Comment by draffelt** [comment]
>  You could have a .row() and an equivalent .save_row() or something method, I suppose.

I was thinking we would use `set_row()` and `get_row()`, but I'm happy with the proxy class option also.

**Comment by Lestropie** [comment]
Just hit this snag again; not the negative strides, but use of `.row()` [within registration code](https://github.com/MRtrix3/mrtrix3/blob/master/lib/filter/warp.h#L37), which forces use of `.with_direct_io()` in a command even though I'm simply looping through the image data once. So going to need to look at this sooner rather than later.

**Comment by jdtournier** [comment]
OK, I just pushed some initial work I'd started on this a while back, see branch [new_row_method](https://github.com/MRtrix3/mrtrix3/tree/new_row_method). This is still very much works in progress, since I'd also mixed in some [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) changes to simplify the process of providing this kind of generic image functionality to e.g. Adapters. This allows us to bring in Value, Index and Row Image::Helpers to these classes relatively easily. I'm also hoping this will help simplify some of the SFINAE stuff, by having all ImageTypes derive from Image::Base (e.g. by removing the need for [this hack](https://github.com/MRtrix3/mrtrix3/blob/new_row_method/lib/image_helpers.h#L142-L164))... 

So far, I've only used it for [`sh2amp`](https://github.com/MRtrix3/mrtrix3/blob/new_row_method/cmd/sh2amp.cpp#L72-L76), and that seems to work fine. But nothing else will built at the moment, these changes will need to be propagated through the rest of the codebase, and I also need to update the Adapters, etc. to derive from ImageBase. But at least it gives you a flavour of a possible solution to this, let me know what you guys think.

**Comment by draffelt** [comment]
Just had a quick look at the branch. Wondering why you recently ditched the CRTP changes?

**Comment by jdtournier** [comment]
> Wondering why you recently ditched the CRTP changes?

Just to keep the history clean, really. I didn't want to introduce unrelated changes all in the same commit or branch. And getting CRTP to work will involve a fair bit of refactoring and checking things over, I'm not sure I have the time to do this right now...

More importantly though: I'd started migrating various parts of the code over, but when I started working over the registration code (particularly the warp handling, e.g. [`src/registration/warp/compose.h`](https://github.com/MRtrix3/mrtrix3/blob/master/src/registration/warp/compose.h#L63)), I started wondering whether this might require a different approach: the code is littered with `.row()` calls in places where it's hard to get rid of cleanly... [This line](https://github.com/MRtrix3/mrtrix3/blob/master/src/registration/warp/compose.h#L63) is an example of where this become unwieldy:
```C++
deform_output.row(3) = linear_transform * (image_transform.voxel2scanner * voxel + disp_input.row(3));
```
With current syntax, this would need to be written as:
```C++
Eigen::Vector3 in;
in << disp_input.row(3);
deform_output.row(3) << linear_transform * (image_transform.voxel2scanner * voxel + in);
```
This is an isolated example, but the code is full of these use cases. Sure, we could just modify everything, but it's a shame to obscure the code like this. So I've been thinking about other possibilities, but I have to admit it's proving very difficult to think of anything else... However, there might be an alternative approach that could somehow address _most_ of the issue - assuming it can be made to work:

Rather than using overloaded extraction/insertion operators, we might actually be able to define proper constructors & assignment operators to do the same job. It's not a problem to add an assignment operator for the Row class, this would look like this:
```C++
template <class ImageType> 
  class Row {
    ...
    template <class Derived>
      void operator= (const Eigen::MatrixBase<Derived>& mat); 
  };
```
What's tricky is the assignment _to_ an Eigen object. But it turns out that Eigen provides a [plugin system](http://eigen.tuxfamily.org/dox/TopicCustomizing_Plugins.html) to add methods to the relevant classes. So we should be able to at least provide an assignment operator with this signature:
```C++
template <class ImageType>
  void operator= (const Row<ImageType>& row);
```
This on its own wouldn't buy us much other than slightly more sane syntax. That previous line would now look like this:
```C++
Eigen::Vector3 in;
in = disp_input.row(3);
deform_output.row(3) = linear_transform * (image_transform.voxel2scanner * voxel + in);
```
Where this would be interesting is if the plugin system allows us to add _constructors_, something with this signature:
```C++
template <class ImageType>
  MatrixBase (const Row& row);
```
If that works, then this might help simplify the syntax somewhat. For example, the line above could then be written as:
```C++
deform_output.row(3) = linear_transform * (image_transform.voxel2scanner * voxel + Eigen::Vector3(disp_input.row(3)));
```
In other words, the only requirement here would be an explicit cast to `Eigen::Vector3` for the row that's being used in an Eigen expression. In this instance, it doesn't add any overhead, since there's no allocation for this particular class. I'll give this a shot in a minute, just thought I'd give you a run-down of where I am with my thought process here... What do you think?

**Comment by draffelt** [comment]
I agree that the overloaded extraction/insertion operators are clunky. Would be great it if you can get the alternative to work

**Comment by jdtournier** [comment]
OK, had a go at this, see #882...
